lib/Sema/NameBinding.cpp

```diff
@@ -36,60 +37,539 @@
+    // Can't process further. checkModuleLoaded() will have diagnosed this.
+    if (I.ID)
+      I.ID.get()->setModule(nullptr);
+    return;
+  }
+
+  auto topLevelModule = I.getTopLevelModule(M, SF);
+
+  I.validateOptions(topLevelModule, SF);
+
+  if (topLevelModule && topLevelModule != M) {
+    // If we have distinct submodule and top-level module, add both, disabling
+    // the top-level module's scoped import validation.
+    addImport(I, M, false);
+    addImport(I, topLevelModule.get(), true);
+  }
+  else {
+    // Add only the import itself.
+    addImport(I, M, true);
+  }
+
+  if (I.ID)
+    I.ID.get()->setModule(M);
+}
+
+void NameBinder::addImport(const UnboundImport &I, ModuleDecl *M,
+                           bool needsScopeValidation) {
+  auto importDesc = I.makeDesc(M);
+
+  unvalidatedImports.emplace_back(I, importDesc, M, needsScopeValidation);
+}
+
+//===----------------------------------------------------------------------===//
+// MARK: Import module loading
+//===----------------------------------------------------------------------===//
+
+ModuleDecl *NameBinder::getModule(ArrayRef<Located<Identifier>> modulePath) {
   assert(!modulePath.empty());
   auto moduleID = modulePath[0];
-  
+
   // The Builtin module cannot be explicitly imported unless we're a .sil file
   // or in the REPL.
   if ((SF.Kind == SourceFileKind::SIL || SF.Kind == SourceFileKind::REPL) &&
-      moduleID.Item == Context.TheBuiltinModule->getName())
-    return Context.TheBuiltinModule;
+      moduleID.Item == ctx.TheBuiltinModule->getName())
+    return ctx.TheBuiltinModule;
 
   // If the imported module name is the same as the current module,
   // skip the Swift module loader and use the Clang module loader instead.
   // This allows a Swift module to extend a Clang module of the same name.
   //
   // FIXME: We'd like to only use this in SIL mode, but unfortunately we use it
-  // for our fake overlays as well.
+  // for clang overlays as well.
   if (moduleID.Item == SF.getParentModule()->getName() &&
       modulePath.size() == 1) {
-    if (auto importer = Context.getClangModuleLoader())
+    if (auto importer = ctx.getClangModuleLoader())
       return importer->loadModule(moduleID.Loc, modulePath);
     return nullptr;
   }
-  
-  return Context.getModule(modulePath);
+
+  return ctx.getModule(modulePath);
+}
+
+NullablePtr<ModuleDecl>
+UnboundImport::getTopLevelModule(ModuleDecl *M, SourceFile &SF) {
+  if (modulePath.size() == 1)
+    return M;
+
+  // If we imported a submodule, import the top-level module as well.
+  Identifier topLevelName = modulePath.front().Item;
+  ModuleDecl *topLevelModule = SF.getASTContext().getLoadedModule(topLevelName);
+
+  if (!topLevelModule) {
+    // Clang can sometimes import top-level modules as if they were
+    // submodules.
+    assert(!M->getFiles().empty() &&
+           isa<ClangModuleUnit>(M->getFiles().front()));
+    return M;
+  }
+
+  if (topLevelModule == SF.getParentModule())
```

Comment:
Wrt this check and the `return topLevelModule` below: at this point in the code, is it not true that `topLevelModule == nullptr`?

Reply:
No—the previous condition is `if (!topLevelModule)`, so we only reach this code if `topLevelModule` is not null.

(`!` is a tiny operator, isn't it?)